### PR TITLE
Feature/database implement

### PR DESCRIPTION
**Pull request by LukaszTheLittle**

### CODE REVIEW

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt [hunk 1]
```diff
@@ -9,28 +9,26 @@ import android.widget.ImageView
...
 import java.util.*
+import androidx.lifecycle.Observer
 
 private const val TAG = "CrimeListFragment"
 
 class CrimeListFragment: Fragment() {
 
     private lateinit var crimeRecyclerView: RecyclerView
-    private var adapter: CrimeAdapter? = null
+    private var adapter: CrimeAdapter? = CrimeAdapter(emptyList())
```
Comment (Pantkowsky) [inline app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt:24]: Why do you need the `adapter` to be a nullable type, if you instantiate it directly and never nullify it? Wouldn't it be better to write
```
private val adapter: CrimeAdapter = CrimeAdapter(emptyList())
```

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt [hunk 2]
```diff
@@ -101,10 +111,11 @@ class CrimeListFragment: Fragment() {
...
 
-    private fun updateUI() {
-        val crimes = crimeListViewModel.crimes
+    private fun updateUI(crimes: List<Crime>) {
         adapter = CrimeAdapter(crimes)
         crimeRecyclerView.adapter = adapter
     }
+}
+
+
```
Comment (Pantkowsky) [inline app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt]: You have two empty lines at the end of file here. By default code style guidelines, there should be a single empty line at the end of file

File: app/src/main/java/database/CrimeDao.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.*
+
+@Dao
+interface CrimeDao {
+
+    @Query("SELECT * FROM crime")
+    fun getCrimes(): LiveData<List<Crime>>
+
+    @Query("SELECT * FROM crime WHERE id=(:id)")
+    fun getCrime(id: UUID): LiveData<Crime?>
```
Comment (Pantkowsky) [inline app/src/main/java/database/CrimeDao.kt:16]: *nit pick*

If you provide an `id` as an argument here, then for the sake of consistency you could consider renaming this method (and any other that ahs the signature of `getCrime(id: String)`) to `getCrimeById`

File: app/src/main/java/database/CrimeTypeConverters.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package database
+
+import androidx.room.TypeConverter
+import java.util.*
```
Comment (Pantkowsky) [inline app/src/main/java/database/CrimeTypeConverters.kt]: By default, it is a [bad practice](https://softwareengineering.stackexchange.com/questions/375901/is-importing-wildcards-a-bad-idea) to use _[wildcard imports](https://tharakamd-12.medium.com/is-it-bad-to-use-wildcard-imports-in-java-1b46a863b2be#:~:text=Wildcard%20imports%20tell%20java%20compiler,performance%20may%20lower%20a%20bit.)_. Check[ this link](https://stackoverflow.com/questions/67341957/how-do-i-disable-wildcard-imports-for-all-kotlin-files-in-android-studio) to see how you can prevent your IDE from doing it.

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt [hunk 5]
```diff
@@ -41,12 +39,24 @@ class CrimeListFragment: Fragment() {
...
 
         return view
     }
 
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        crimeListViewModel.crimeListLiveData.observe(
+            viewLifecycleOwner,
+            Observer { crimes ->
+                crimes?.let {
```
Comment (Pantkowsky) [inline app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt:50]: Do you need to check for nullability of `crimes` in line `52`, i.e 
```
crimes?.let { ... }
```

Isn't `crimes` already `non-null`?

**Review by Pantkowsky — CHANGES_REQUESTED** [review]
Cool stuff ! 👍 

There are a couple of comments you should address though. Two things to keep an eye out here:
- [null safety and safe calls](https://kotlinlang.org/docs/null-safety.html#nullable-types-and-non-null-types)
- wildcard imports

**Comment by LukaszTheLittle** [comment]
All changes suggested by @Pantkowsky have been made

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt (hunk 2 above)
Comment (LukaszTheLittle) [reply app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt]: Unnecessary line deleted

File: app/src/main/java/database/CrimeDao.kt (hunk 3 above)
Comment (LukaszTheLittle) [reply app/src/main/java/database/CrimeDao.kt:16]: Will keep it in mind

File: app/src/main/java/database/CrimeTypeConverters.kt (hunk 4 above)
Comment (LukaszTheLittle) [reply app/src/main/java/database/CrimeTypeConverters.kt]: Didn't know that. Thanks a lot. All wildcard imports have been deleted across all files.

**Pull request merged**